# Objetivo del proceso a realizar				
## Objetivos del desafío 1			
- Efectuar una limpieza del dataset provisto. Particularmente, deberá diseñar estrategias para lidiar con los datos perdidos en ciertas variables.		
- Realizar un análisis descriptivo de las principales variables.		
- Crear nuevas columnas a partir de las características dadas que puedan tener valor predictivo.		

### 1) Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
%matplotlib inline

In [2]:
data = pd.read_csv('properatti.csv', sep = ",")

In [3]:
df = data.copy()

### 2) Analisis General del Data Set

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121220 entries, 0 to 121219
Data columns (total 26 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Unnamed: 0                  121220 non-null  int64  
 1   operation                   121220 non-null  object 
 2   property_type               121220 non-null  object 
 3   place_name                  121197 non-null  object 
 4   place_with_parent_names     121220 non-null  object 
 5   country_name                121220 non-null  object 
 6   state_name                  121220 non-null  object 
 7   geonames_id                 102503 non-null  float64
 8   lat-lon                     69670 non-null   object 
 9   lat                         69670 non-null   float64
 10  lon                         69670 non-null   float64
 11  price                       100810 non-null  float64
 12  currency                    100809 non-null  object 
 13  price_aprox_lo

Revisamos el porcentaje de nulos

In [5]:
df.isnull().sum()/df.shape[0]*100

Unnamed: 0                     0.000000
operation                      0.000000
property_type                  0.000000
place_name                     0.018974
place_with_parent_names        0.000000
country_name                   0.000000
state_name                     0.000000
geonames_id                   15.440521
lat-lon                       42.525986
lat                           42.525986
lon                           42.525986
price                         16.837156
currency                      16.837981
price_aprox_local_currency    16.837156
price_aprox_usd               16.837156
surface_total_in_m2           32.443491
surface_covered_in_m2         16.422208
price_usd_per_m2              43.394654
price_per_m2                  27.686850
floor                         93.483749
rooms                         60.905791
expenses                      88.234615
properati_url                  0.000000
description                    0.001650
title                          0.000000


### 3) Analisis de duplicados

Vamos a ver si tenemos inmuebles duplicados en nuetro dataset.

Definimos que tenemos un duplicado segun la columna 'properati_url', ya que 2 publicaciones distintas, no pueden remitir a la misma URL de properati.

In [6]:
df.duplicated('properati_url').any()

False

No hay filas duplicadas

Verificamos que todas las propiedades provistas estan para la venta, y ninguna en renta, con lo cual la columna 'operation' no nos aporta mucha informacion.

In [7]:
df.operation.value_counts()

sell    121220
Name: operation, dtype: int64

Vamos a revisar los valores estadisticos de las variables numericas de nuestro Data set

In [8]:
cols=['price','price_aprox_local_currency','price_aprox_usd','surface_total_in_m2','surface_covered_in_m2','price_usd_per_m2','price_per_m2']
df[cols].describe()

price  price_aprox_local_currency  price_aprox_usd  \
count  1.008100e+05                1.008100e+05     1.008100e+05   
mean   4.685259e+05                4.229397e+06     2.397006e+05   
std    2.260101e+06                6.904714e+06     3.913239e+05   
min    0.000000e+00                0.000000e+00     0.000000e+00   
25%    1.100000e+05                1.583309e+06     8.973388e+04   
50%    1.850000e+05                2.558452e+06     1.450000e+05   
75%    4.200000e+05                4.675792e+06     2.650000e+05   
max    6.500000e+08                8.212711e+08     4.654544e+07   

       surface_total_in_m2  surface_covered_in_m2  price_usd_per_m2  \
count         81892.000000          101313.000000      68617.000000   
mean            233.795328             133.050181       2160.086916   
std            1782.222147             724.351479       2759.288621   
min               0.000000               0.000000          0.600000   
25%              50.000000              45.000000       1218.181818   
50%              84.000000              75.000000       1800.000000   
75%             200.000000             150.000000       2486.411765   
max          200000.000000          187000.000000     206333.333333   

       price_per_m2  
count  8.765800e+04  
mean   6.912216e+03  
std    2.837864e+04  
min    1.510204e+00  
25%    1.550000e+03  
50%    2.213115e+03  
75%    3.355549e+03  
max    4.000000e+06

Aca podemos observar algunos valores extraños en el data set:

* El valor minimo de superficie cubierta y total, es 0.

* El valor minimo de venta es 0.

* El valor maximo de superficie llega a 20 hectareas


### 4) Analisis de columnas

**Analisis Descriptivo de las principales variables:**

* price_per_m2: Es nuestra variable objetivo, es la que vamos a querer predecir en la siguiente etapa del trabajo, nos representa el precio por metro cuadrado del inmueble.VARIABLE CONTINUA

* price_usd_per_m2: Esta variable nos indica el valor en dolares por metro cuadrado del inmueble publicado.VARIABLE CONTINUA

* price: Esta variable nos indica el precio del inmueble expresado en la moneda "currency".VARIABLE CONTINUA

* currency: Esta variable nos indica la moneda en la que se publico el inmueble, puede valer (usd,ars,uyu,pen).VARIABLE CATEGORICA

* surface_total_in_m2: Esta variable nos indica los metros cuadrados totales de la propiedad publicada.VARIABLE CONTINUA

* surface_covered_in_m2: Esta variable nos indica los metros cuadrados cubiertos de la propiedad publicada.VARIABLE CONTINUA

* property_type: Nos indica el tipo de propiedad publicada (ph, apartment, house, store).VARIABLE CATEGORICA

* place_name: Nos indica el barrio(si es dentro de CABA) o localidad, de donde esta la propiedad publicada.VARIABLE CATEGORICA

* state_name: Nos indica la provincia en donde se encuentra la propiedad publicada.VARIABLE CATEGORICA

* place_with_parent_names: Es una combinacion de place_name y state_name, en algunos casos, con algunos datos mas sobre la ubicacion, como el partido. place_with_parent_names = |country_name|state_name|place_name|.VARIABLE CATEGORICA

* floor: Numero de piso en donde esta ubicado el inmueble(en caso de corresponder).VARIABLE CUALITATIVA ORDINAL

* rooms: Esta variable nos indica la cantidad de ambientes que posee el inmueble.VARIABLE CATEGORICA

* expenses: Valor de las expensas del inmueble (en caso de corresponder).VARIABLE CONTINUA

* description: Una breve descripcion sobre la propiedad en venta.

In [9]:
#Para evitar problemas con los strings, convertimos todo el df en minuscula
df = df.applymap(lambda x: x.lower() if type(x) is str else x)

### 5) Análisis inicial de la distribución de los datos

Separamos las columnas que consideramos tengan alguna relacion con el precio_per_m2 o nos puedan brindar informacion util 

In [10]:
cols_utiles=['property_type','place_name','place_with_parent_names','state_name','price','currency'\
            ,'price_aprox_local_currency','price_aprox_usd','surface_total_in_m2','surface_covered_in_m2'\
            ,'price_usd_per_m2','price_per_m2','floor','rooms','expenses','description']
cols_utiles_sin_price=['property_type','place_name','place_with_parent_names','state_name','currency'\
            ,'price_aprox_local_currency','price_aprox_usd','surface_total_in_m2','surface_covered_in_m2'\
            ,'price_usd_per_m2','price_per_m2','floor','rooms','expenses','description']

In [11]:
groupby_property_type=df.groupby('property_type')[cols_utiles_sin_price].median()
groupby_property_type

price_aprox_local_currency  price_aprox_usd  \
property_type                                                
apartment                       2099695.5         119000.0   
house                           4234680.0         240000.0   
ph                              2117340.0         120000.0   
store                           3705345.0         210000.0   

               surface_total_in_m2  surface_covered_in_m2  price_usd_per_m2  \
property_type                                                                 
apartment                     59.0                   52.0       2167.832168   
house                        278.0                  180.0       1057.934509   
ph                            90.0                   73.0       1328.619462   
store                        154.0                  120.0       1637.931034   

               price_per_m2  floor  rooms  expenses  
property_type                                        
apartment       2710.072464    3.0    2.0    1700.0  
house           1529.411765    1.0    4.0    4500.0  
ph              1718.750000    2.0    3.0     500.0  
store           2100.000000    1.0    1.0    1111.0

Dado que 'price', depende del valor de currency, si esta en usd o en ars, vamos a calcularlos por separados, para no crear un bias.

In [12]:
groupby_property_type_price_usd=df[df.currency=='usd'].groupby('property_type')['price'].median()
groupby_property_type_price_ars=df[df.currency=='ars'].groupby('property_type')['price'].median()

groupby_property_type['price_usd']=groupby_property_type_price_usd
groupby_property_type['price_ars']=groupby_property_type_price_ars
groupby_property_type['cant de propiedades']=df.groupby('property_type')['operation'].count()

groupby_property_type

price_aprox_local_currency  price_aprox_usd  \
property_type                                                
apartment                       2099695.5         119000.0   
house                           4234680.0         240000.0   
ph                              2117340.0         120000.0   
store                           3705345.0         210000.0   

               surface_total_in_m2  surface_covered_in_m2  price_usd_per_m2  \
property_type                                                                 
apartment                     59.0                   52.0       2167.832168   
house                        278.0                  180.0       1057.934509   
ph                            90.0                   73.0       1328.619462   
store                        154.0                  120.0       1637.931034   

               price_per_m2  floor  rooms  expenses  price_usd  price_ars  \
property_type                                                               
apartment       2710.072464    3.0    2.0    1700.0   129394.0  1420000.0   
house           1529.411765    1.0    4.0    4500.0   250000.0  1450000.0   
ph              1718.750000    2.0    3.0     500.0   125000.0  1350000.0   
store           2100.000000    1.0    1.0    1111.0   235000.0  1100000.0   

               cant de propiedades  
property_type                       
apartment                    71065  
house                        40268  
ph                            5751  
store                         4136

Al agrupar las propiedades por su tipo, podemos realizar las siguientes observaciones:

*  Las casas son las que suelen tener un mayor precio, pero tambien son las que tienen mas metros cuadrados tienen y pagan mas expensas.

*  Los departamentos son las propiedades que mayor valor por metro cuadrado tienen.

*  Los ph, son la opcion mas economica de propiedad, ya que tienen el precio mas bajo y el menor valor de expensas.

*  La mayoria de los inmuebles publicados son departamentos.

In [13]:
groupby_provincia_price_usd=df[df.currency=='usd'].groupby('state_name')['price'].median()
groupby_provincia_price_ars=df[df.currency=='ars'].groupby('state_name')['price'].median()

groupby_provincia=df.groupby('state_name')[cols_utiles_sin_price].median()

groupby_provincia['price_usd']=groupby_provincia_price_usd
groupby_provincia['price_ars']=groupby_provincia_price_ars
groupby_provincia['cant de propiedades']=df.groupby('state_name')[cols_utiles_sin_price].count().property_type


groupby_provincia.sort_values('cant de propiedades',ascending=False)

price_aprox_local_currency  price_aprox_usd  \
state_name                                                                  
capital federal                             2.774792e+06       157261.000   
bs.as. g.b.a. zona norte                    4.058235e+06       230000.000   
bs.as. g.b.a. zona sur                      2.470230e+06       140000.000   
córdoba                                     1.729152e+06        97999.500   
santa fe                                    1.405187e+06        79638.820   
buenos aires costa atlántica                1.905606e+06       108000.000   
bs.as. g.b.a. zona oeste                    1.940895e+06       110000.000   
buenos aires interior                       2.382008e+06       135000.000   
río negro                                   2.734898e+06       155000.000   
neuquén                                     2.180514e+06       123580.375   
mendoza                                     2.196740e+06       124500.000   
tucumán                                     1.583309e+06        89733.880   
corrientes                                  2.693433e+06       152650.000   
misiones                                    1.499782e+06        85000.000   
entre ríos                                  1.682266e+06        95342.250   
salta                                       1.710715e+06        96954.570   
chubut                                      3.528900e+06       200000.000   
san luis                                    2.205562e+06       125000.000   
la pampa                                    1.632788e+06        92538.065   
formosa                                     8.814045e+05        49953.500   
chaco                                       2.473921e+06       140209.190   
san juan                                    5.046799e+06       286026.750   
tierra del fuego                            2.177050e+06       123384.080   
catamarca                                   1.076314e+07       610000.000   
jujuy                                       2.464025e+06       139648.350   
santa cruz                                  2.473921e+06       140209.190   
la rioja                                    1.075510e+06        60954.385   
santiago del estero                         3.176010e+06       180000.000   

                              surface_total_in_m2  surface_covered_in_m2  \
state_name                                                                 
capital federal                              62.0                   55.0   
bs.as. g.b.a. zona norte                    160.0                  120.0   
bs.as. g.b.a. zona sur                      110.0                   94.0   
córdoba                                      79.0                   85.0   
santa fe                                     60.0                   53.0   
buenos aires costa atlántica                 75.0                   70.0   
bs.as. g.b.a. zona oeste                    125.0                   80.0   
buenos aires interior                       220.0                  110.0   
río negro                                   150.0                  105.0   
neuquén                                      70.0                   63.0   
mendoza                                     118.0                  100.0   
tucumán                                     100.0                   90.0   
corrientes                                  110.0                  120.0   
misiones                                    115.0                   84.0   
entre ríos                                  134.5                   95.0   
salta                                       150.0                   94.5   
chubut                                      134.0                  124.0   
san luis                                    240.0                  120.0   
la pampa                                    125.0                  110.0   
formosa                                      40.0                    NaN   
chaco                                       200

Al agrupar las propiedades por provincia, podemos realizar las siguientes observaciones:

*  Las propiedades que se encuentran en capital federal, son las que tienen un precio por metro cuadrado en usd mas alto.

*  Las propiedades que se encuentran en jujuy, son las que tienen mayor cantidad de metros cuadrados.

*  La mayor cantidad de propiedades en venta, se encuentran en capital federal.


### 6) Analisis de consistencia y calidad de los datos

#### Los precios estan en distintas monedas, vamos a llevar todo a usd.

In [14]:
df.currency.value_counts()

usd    87587
ars    13219
pen        2
uyu        1
Name: currency, dtype: int64

##### Primero, vamos a estudiar los casos donde currency vale pen y uyu

In [15]:
df_currency_uyu=df[df.currency=='uyu']
list(df_currency_uyu.description)

['casa con calefacción central, mendoza, san roque, por $ 650.000.000']

In [16]:
df_currency_uyu

Unnamed: 0 operation property_type place_name  \
107390      107390      sell         house  san roque   

              place_with_parent_names country_name state_name  geonames_id  \
107390  |argentina|mendoza|san roque|    argentina    mendoza    3836586.0   

       lat-lon  lat  ...  surface_covered_in_m2  price_usd_per_m2  \
107390     NaN  NaN  ...                    NaN               NaN   

       price_per_m2  floor  rooms  expenses  \
107390          NaN    NaN    NaN       NaN   

                                            properati_url  \
107390  http://www.properati.com.ar/1by0s_venta_casa_s...   

                                              description               title  \
107390  casa con calefacción central, mendoza, san roq...  casa - san vicente   

                                          image_thumbnail  
107390  https://thumbs4.properati.com/2/pa0rbudkrk6ggv...  

[1 rows x 26 columns]

In [17]:
#Entonces vamos a imputarle el valor de price = $ 650.000.000, con currency='ars'
df.loc[107390,'price']=650000000
df.loc[107390,'currency']='ars'
df.loc[107390,['price','currency']]

price       650000000.0
currency            ars
Name: 107390, dtype: object

In [18]:
list(df[df.currency=='pen'].description)

['ref.: 55:preventa ::: oportunidad ::: dúplex a entregarse en 20 meses de 47m2, pequeño patio. con todos los artefactos.- departamento de 1 dormitorios con patio, cochera y parrilla entrega: $380.000 + 24 cuotas de $24.000 ( ) departamento de 1 dormitorio con patio, sin cochera entrega: $350.000 + 24 cuotas de $22.000 (  ) (  ) cuotas ajustadas al indice de la construcción.-  ',
 'ref.: 56duplex en venta de 1 dormitorio con aproximadamente 50m2 c/u hay 3 departamentos en venta, todos con cocina, termotanque y calefactores. ']

En este caso, para el primer inmueble, tomamos el valor con cochera y nos daria un total de $1.239.063,536, con currency='ars', y tomando para las cuotas un indice CAC = 19.10% y 50.45% para 2017 y 2018. Tambien asumimos que el valor 47m2 refiere a la superficie cubierta en metros cuadrados


https://ikiwi.net.ar/indice-cac/#Expandocac2017 

In [19]:
df.loc[50387,'price']=12390635.36
df.loc[50387,'currency']='ars'
df.loc[50387,'surface_covered_in_m2']=47

df.loc[50387,['price','currency','surface_covered_in_m2']]

price                    12390635.36
currency                         ars
surface_covered_in_m2           47.0
Name: 50387, dtype: object

El segundo caso, es un poco confusa la descripcion, asi que tomamos el criterio de dropearlo.

In [20]:
df[df.currency=='pen']

Unnamed: 0 operation property_type       place_name  \
50388       50388      sell     apartment  capital federal   

           place_with_parent_names country_name       state_name  geonames_id  \
50388  |argentina|capital federal|    argentina  capital federal    3433955.0   

      lat-lon  lat  ...  surface_covered_in_m2  price_usd_per_m2 price_per_m2  \
50388     NaN  NaN  ...                    NaN               NaN          NaN   

       floor  rooms  expenses  \
50388    NaN    NaN       NaN   

                                           properati_url  \
50388  http://www.properati.com.ar/19e90_venta_depart...   

                                             description  \
50388  ref.: 56duplex en venta de 1 dormitorio con ap...   

                                           title  \
50388  duplex a la venta | liniers 1200 | tandil   

                                         image_thumbnail  
50388  https://thumbs4.properati.com/8/_cbxayr_9pdysg...  

[1 rows x 26 columns]

In [21]:
df.drop(index=50388, inplace=True)
df.reset_index(drop=True,inplace=True)

In [ ]:
### ACA ME QUEDEEEEEEEEEEEEE, SEGUIRRRRR, VIENDO SI PUEDO TRANSFORMAR TODO A USD O TENGO QUE SEGUIR TRABAJANDO EN ARS Y USD

aaa

In [24]:
df[cols_utiles].head(1)

property_type place_name                place_with_parent_names  \
0            ph  mataderos  |argentina|capital federal|mataderos|   

        state_name    price currency  price_aprox_local_currency  \
0  capital federal  62000.0      usd                   1093959.0   

   price_aprox_usd  surface_total_in_m2  surface_covered_in_m2  \
0          62000.0                 55.0                   40.0   

   price_usd_per_m2  price_per_m2  floor  rooms  expenses  \
0       1127.272727        1550.0    NaN    NaN       NaN   

                                         description  
0  2 ambientes tipo casa planta baja por pasillo,...

In [ ]:
price_per_m2:
    currency=usd
        price_per_m2 = price/surface_covered_in_m2
        Si coinciden surface_covered_in_m2 y surface_total_in_m2, entonces price_per_m2 == price_usd_per_m2
    currency=ars
        price_per_m2=price/surface_covered_in_m2(apartment and house), hay que chequear con store
        Si coinciden surface_covered_in_m2 y surface_total_in_m2, entonces price_usd_per_m2 == 17.830501*price_per_m2
    currency=uyu
        segun la descripcion, hay q ponerle price=650.000.000 y currency=ars
    currency=pen
        Segun la descripcion es un departamento de pozo, con cochera, lo vamos a tomar $956.000, currency=ars, 
        tomamos la superficie que nos dice como surface_covered_in_m2=47m2.

### Floor

Sabemos que los departamentos que se encuentran en los pisos superiores de los edificios, suelen tener precios superiores a los que se encuentran cerca de la planta baja, pero nuestra columna floor, esta muy vacia, vamos a ver si podemos completarla con la informacion que encontramos en la columna 'description'

In [ ]:
df.property_type.value_counts()

In [ ]:
departamento_mask = df.property_type=='apartment'
patron_piso = re.compile(r'(\d+(º)?(\s)?(piso))|((piso)(\s)?\d+(º)?)')

encuentra_piso=df.loc[departamento_mask,'description'].apply(lambda x: re.search(patron_piso,str(x)))

encuentra_piso_mask=encuentra_piso.notnull()

pisos=encuentra_piso[encuentra_piso_mask].apply(lambda x: x.group())


In [ ]:
# pisos.shape
pisos.sample(10)

Nuestro REGEX matchea dentro de todo bastante bien, solo debemos filtrar algunos valores altos. Pero primero, vamos a filtrar y convertir a int, nuestra variable pisos

In [ ]:
pisos_sin_piso=pisos.replace("piso"," ",regex=True)
pisos_sin_piso = pisos_sin_piso.apply(lambda x: str(x).strip())
pisos_clean_str = pisos_sin_piso.replace("º","",regex=True)
pisos_int = pisos_clean_str.apply(lambda y: int(y))
pisos_int.sample(5)

Ahora, voy a intentar sacarme de encima a los pocos valores que el regex me tomo por error, primero defino con mi criterio, que los edificios no pueden tener mas de 40 pisos

In [ ]:
pisos_filtrado_mask = pisos_int <= 50

pisos_filtrado_menor50=pisos_int[pisos_filtrado_mask]

In [ ]:
g=sns.histplot(pisos_filtrado_menor50)
sns.set(rc = {'figure.figsize':(15,6)})
g.set_title("DISTRIBUCION DE LOS DEPARTAMENTOS SEGUN LOS PISOS", fontsize = 17)
g.set_xlabel("PISO DE LA PROPIEDAD", fontsize = 15)
g.set_ylabel("CANTIDAD DE DEPARTAMENTOS", fontsize = 15)

La gran mayoria de los departamentos, rondan entre el 2do y 10mo piso

In [ ]:
pisos_filtrado_menor50.median()

Porcentaje de valores faltantes en floor para viviendas tipo departamento

In [ ]:
df.floor[departamento_mask].isnull().sum()/df.floor[departamento_mask].shape[0]*100

Ahora con estos valores de pisos, voy a imputar mi columna floor, en las filas donde property_type sea apartment y floor tenga valor nan

In [ ]:
null_floor_mask = df.floor.isnull()
departamento_mask
departamento_y_floor_null=np.logical_and(departamento_mask,null_floor_mask)

df.loc[departamento_y_floor_null,'floor']=pisos_filtrado_menor50

In [ ]:
df.floor[departamento_mask].isnull().sum()/df.floor[departamento_mask].shape[0]*100

Bajo el porcentaje de nulos, pero aun sigue siendo muy alto para tomar esta columna como relevante.

### Rooms

Es posible que la cantidad de ambientes en los departamentos, que se encuentren en la misma zona (place_name), influya en el valor del inmueble, pero tenemos un 60.9% de columnas Null, veremos si podemos obtener alguna informacion de la columna description, como hicimos con floor.

In [ ]:
patron_ambientes = re.compile(r'\d+(\s)?(ambiente(s)?)|\d+(\s)?(amb(\.)?)')


encuentra_ambientes=df.loc[departamento_mask,'description'].apply(lambda x: re.search(patron_ambientes,str(x)))

encuentra_ambientes_mask=encuentra_ambientes.notnull()

ambientes=encuentra_ambientes[encuentra_ambientes_mask].apply(lambda x: x.group())
ambientes.sample(30)
# ambientes.shape
ambientes.value_counts()

Podemos ver que nuestro Regex funciono bastante bien, pero hay alguno valores que hay que filtrar, para esto primero vamos a convertir a dato numerico 

In [ ]:
ambientes_sin_ambientes=ambientes.replace("ambientes"," ",regex=True)
ambientes_sin_ambientes_sin_ambiente=ambientes_sin_ambientes.replace("ambiente"," ",regex=True)
ambientes_sin_ambientes_sin_ambiente_sin_amb=ambientes_sin_ambientes_sin_ambiente.replace("amb(\.)?"," ",regex=True)

ambientes_clean_str = ambientes_sin_ambientes_sin_ambiente_sin_amb.apply(lambda x: str(x).strip())
ambientes_int = ambientes_clean_str.apply(lambda y: int(y))
ambientes_int.sample(5)

Voy a tomar un criterio de que mas de 10 ambientes, en un departamento, es un valor incorrecto

In [ ]:
ambientes_menor_10=ambientes_int<10
ambientes_clean=ambientes_int[ambientes_menor_10]

In [ ]:
g1=sns.histplot(ambientes_clean)
sns.set(rc = {'figure.figsize':(15,6)})
g1.set_title("CANTIDAD DE DEPARTAMENTOS EN FUNCION DE SU N DE AMBIENTES", fontsize = 17)
g1.set_xlabel("Nº AMBIENTES", fontsize = 15)
g1.set_ylabel("CANTIDAD DE DEPARTAMENTOS", fontsize = 15)

Podemos observar que la gran mayoria de los departamentos en venta, tienen 2 y 3 ambientes

Porcentaje de valores faltantes en la columna rooms, para viviendas tipo departamento.

In [ ]:
df.rooms[departamento_mask].isnull().sum()/df.rooms[departamento_mask].shape[0]*100

Ahora vamos a imputar con ambientes_clean, algunos de los valores faltantes de rooms

In [ ]:
rooms_null_mask=df.rooms.isnull()
departamento_rooms_null_mask = np.logical_and(rooms_null_mask,departamento_mask)

df.loc[departamento_rooms_null_mask,'rooms']=ambientes_clean

df.rooms[departamento_mask].isnull().sum()/df.rooms[departamento_mask].shape[0]*100

Podemos ver como bajamos casi un 20% el valor de NaN rows en los inmuebles tipo departamento

### Cocheras

Vamos a filtrar los departamentos que tengan cochera, para estudiar si el precio por m2 aumenta en estos casos.

In [ ]:
df_departamentos = df[departamento_mask]

patron_cochera = re.compile(r'cochera(s)?|garage(s)?|garaje(s)?')

encuentra_cochera=df.loc[departamento_mask,'description'].apply(lambda x: re.search(patron_cochera,str(x)))

departamentos_con_cochera_mask=encuentra_cochera.notnull()


departamentos_con_cochera = df_departamentos[departamentos_con_cochera_mask]
departamentos_con_cochera.sample(3)

### Expenses

## Piscinas

Vamos a ver de todas las propiedades, cuales tienen pileta, para ver si esto afecta al valor por metro cuadrado de la propiedad

In [ ]:
patron_pileta = re.compile(r'piscina|pileta')

encuentra_pileta=df['description'].apply(lambda x: re.search(patron_pileta,str(x)))

inmuebles_con_pileta_mask=encuentra_pileta.notnull()


inmuebles_con_pileta = df[inmuebles_con_pileta_mask]

In [ ]:
list(inmuebles_con_pileta.description.sample(1))

## Filtrado de columnas con demasiados elementos faltantes, que puedan producirnos un bias muy alto si queremos imputarlas de alguna manera

Vamos a quedarnos con aquellas columnas que por lo menos tengan el 60% de los valores sin NaN.

In [ ]:
threshold= df.shape[0]*0.6
df_cols_utiles = df[cols_utiles].dropna(thresh = threshold ,axis=1)

In [ ]:
df_cols_utiles.shape[1]

In [ ]:
len(cols_utiles)

De esta manera, dropeamos las columnas 'rooms', 'floor', 'expenses', 'price_usd_per_m2'

In [ ]:
df_cols_utiles.isnull().sum()/df_cols_utiles.shape[0]*100

Dado que 'price_usd_per_m2', está muy cerca del umbral, veremos si tiene alguna utilidad con nuestra variable mas relevante o podemos descartarlo directamente.

In [ ]:
notnull_price_usd_per_m2_mask=df['price_usd_per_m2'].notnull()
null_price_per_m2_mask = df['price_per_m2'].isnull()

(notnull_price_usd_per_m2_mask==null_price_per_m2_mask).any()
#Hay alguna fila donde price_usd_per_m2, tenga un valor util y price_per_m2, sea NaN?

Debido a la proximidad con el umbral que fijamos, y que puede aportarnos informacion util, vamos a cambiar el umbral al 55% de valores utiles, para no perder 'price_usd_per_m2'

In [ ]:
threshold2= df.shape[0]*0.55
df_cols_utiles2 = df[cols_utiles].dropna(thresh = threshold2 ,axis=1)

In [ ]:
df_cols_utiles2.isnull().sum()/df_cols_utiles2.shape[0]*100

In [ ]:
df_cols_utiles2.sample(5)

**Ahora vamos a comprobar si columnas como'price_aprox_local_currency','price_aprox_usd','place_with_parent_names','currency', tienen alguna utilidad para nosotros, o la informacion que nos dan, las tenemos contenidas en otras columnas**

## 'price_aprox_local_currency','price_aprox_usd',

In [ ]:
df_cols_utiles2[['price_aprox_local_currency','price_aprox_usd','price','currency']].sample(5)

Observando el dataset, vemos que price_aprox_local_currency(ars), se aproxima al valor de price, cuando currency esta en 'ars'

Tambien vemos que price_aprox_usd, coincide con el valor de price, cuando currency esta en 'usd'.


In [ ]:
ars_mask = df_cols_utiles2.currency == 'ars'

df_cols_utiles2_ars=df_cols_utiles2[ars_mask]

rel = df_cols_utiles2_ars.price/df_cols_utiles2_ars.price_aprox_local_currency
relacion = rel.mean()
relacion

Observamos que la relacion entre price y price_aprox_local_currency, es "relacion"


Nuestro objetivo, es eliminar price_aprox_local_currency y price_aprox_usd, una vez las usemos para completar price

Veamos primero si nos son utiles estas columnas para lo que queremos hacer.

In [ ]:
notnull_price_aprox_local_currency=df_cols_utiles2.price_aprox_local_currency.notnull()
notnull_price_aprox_usd=df_cols_utiles2.price_aprox_usd.notnull()
price_ars_usd = (df_cols_utiles2.currency == 'usd') | (df_cols_utiles2.currency == 'ars')
null_price_ars_usd=df_cols_utiles2.price[price_ars_usd].isnull()


price_local_o_price_usd=np.logical_or(notnull_price_aprox_local_currency,notnull_price_aprox_usd)

null_price_notnull_price_local_o_price_usd=np.logical_and(price_local_o_price_usd,null_price_ars_usd)

null_price_notnull_price_local_o_price_usd.any()

Por lo tanto, comprobamos que **NO HAY** ningun valor no nulo en price_aprox_usd o price_aprox_local_currency, que podamos utilizar para imputar price, con currency 'ars' o 'usd'

## place_with_parent_names

place_with_parent_names = |country_name|state_name|place_name|, con lo cual, no nos da informacion extra, y es mas facil utilizar para los groupby sus columnas "hijas", con lo cual, la utilizaremos para completar "place_name", que tiene algunos nulos y luego la descartaremos.

Vamos a comprobar si podemos completar place_name, con place_with_parent_names. Para esto, vamos a eliminar las filas donde haya NaN en place_name y comparar si es igual al anteultimo elemento de place_with_parent_names, al aplicarle un split(el ultimo es el pipe).

In [ ]:
place_name_tester=df_cols_utiles2.place_with_parent_names.apply(lambda x: x.split('|')[-2])
place_name_tester

In [ ]:
place_name_null_mask = df_cols_utiles2.place_name.isnull()

In [ ]:
(place_name_tester[~place_name_null_mask]==df_cols_utiles2.place_name[~place_name_null_mask]).all()

Comprobamos que podemos imputar el valor de place_name, a partir de la columna place_with_parent_names, asi que vamos a imputarlo

In [ ]:
df_cols_utiles2.loc[place_name_null_mask,'place_name'] = place_name_tester

df_cols_utiles2.place_name.isnull().sum()

Ahora dropeamos algunas columnas de nuestro data frame

In [ ]:
df_cols_utiles2.head(1)

In [ ]:
df_cols_utiles3=df_cols_utiles2.drop(['place_with_parent_names','price_aprox_local_currency','price_aprox_usd'], axis=1)
df_cols_utiles3.head(3)

Estudiando el dataset, podemos observar, que el valor de price_usd_per_m2, es igual al de price_per_m2, cuando surface_covered_in_m2 es igual a surface_total_in_m2, SIEMPRE QUE CURRENCY ESTE EN USD

In [ ]:
df_cols_utiles2[df_cols_utiles2.surface_covered_in_m2 == df_cols_utiles2.surface_total_in_m2][['price_usd_per_m2','price_per_m2']]

In [ ]:
## IMPUTACION A price_usd_per_m2 con ,price_per_m2 para filas donde currency es 'usd'

In [ ]:
currency_usd_mask=df_cols_utiles2.currency == 'usd'
cubiertos_igual_totales=df_cols_utiles2.surface_covered_in_m2==df_cols_utiles2.surface_total_in_m2

df_price_per_m2_igual_price_usd_per_m2= df_cols_utiles2[np.logical_and(currency_usd_mask,cubiertos_igual_totales)]

df_price_per_m2_igual_price_usd_per_m2.price_per_m2.isnull().any()

Comprobamos que **NO HAY** ningun valor de `price_usd_per_m2` que podamos imputar usando la relacion el valor de `price_per_m2`, es igual al de price_per_m2, cuando surface_covered_in_m2 es igual a surface_total_in_m2, SIEMPRE QUE CURRENCY ESTE EN USD

In [ ]:
df_cols_utiles2.price_usd_per_m2.notnull().sum()

In [ ]:
df_cols_utiles2.price_per_m2.notnull().sum()

In [ ]:
## IMPUTACION A price_per_m2 con price_usd_per_m2, para filas donde currency es 'usd'

In [ ]:
df_cols_utiles2.currency.value_counts()

Para las propiedades en donde currency esta en 'ars', podemos observar que price_per_m2 coincide con price/surface_covered_in_m2


In [ ]:
AAA

In [ ]:
currency_ars_mask=df_cols_utiles2.currency == 'ars'
df_cols_ars = df_cols_utiles2[currency_ars_mask]
df_price_surface_covered = df_cols_ars['price'] / df_cols_ars.surface_covered_in_m2

b = df_cols_utiles2[currency_ars_mask].price_per_m2.dropna()
a = df_price_surface_covered.dropna()

(a==b).all()

In [ ]:
pesos_mask=df_cols_utiles2.currency == 'ars'

df_cols_utiles2[pesos_mask].sample(5)

In [ ]:
uyu_mask=df_cols_utiles2.currency == 'uyu'

df_cols_utiles2[uyu_mask].sample()

In [ ]:
pen_mask=df_cols_utiles2.currency == 'pen'

df_cols_utiles2[pen_mask].sample(2)